File: src/proxy.mjs
import { Socks5ProxyAgent } from "undici";
import { getSocks5Proxy } from "./config.mjs";

let cachedDispatcher = null;

function normalizeProxyUrl(proxyStr) {
  const trimmed = proxyStr.trim();
  if (/^socks5?:\/\//.test(trimmed)) return trimmed;
  return `socks5://${trimmed}`;
}

export function getProxyDispatcher() {
  const socks5Proxy = getSocks5Proxy();
  if (!socks5Proxy) return undefined;
  if (cachedDispatcher) return cachedDispatcher;

  const url = normalizeProxyUrl(socks5Proxy);
  cachedDispatcher = new Socks5ProxyAgent(url);
  return cachedDispatcher;
}

export function proxyFetchOptions(options = {}) {
  const dispatcher = getProxyDispatcher();
  if (!dispatcher) return options;
  return { ...options, dispatcher };
}

export async function proxyFetch(url, options = {}) {
  return fetch(url, proxyFetchOptions(options));
}

export function getProxyInfo() {
  const socks5Proxy = getSocks5Proxy();
  if (!socks5Proxy) return null;
  try {
    const url = new URL(normalizeProxyUrl(socks5Proxy));
    return {
      host: url.hostname,
      port: parseInt(url.port, 10),
      hasAuth: !!(url.username || url.password),
    };
  } catch {
    return { host: socks5Proxy, port: 0, hasAuth: false };
  }
}

File: src/proxy.py
from __future__ import annotations

from .config import get_socks5_proxy

_client = None


def _normalize_url(proxy_str: str) -> str:
    s = proxy_str.strip()
    if s.startswith("socks5://") or s.startswith("socks://"):
        return s
    return f"socks5://{s}"


def get_http_client():
    """Return a shared httpx.AsyncClient with optional SOCKS5 proxy."""
    global _client
    if _client is not None:
        return _client

    import httpx

    proxy = get_socks5_proxy()
    proxy_url = _normalize_url(proxy) if proxy else None
    _client = httpx.AsyncClient(proxy=proxy_url, timeout=60, follow_redirects=True)
    return _client


def get_proxy_info() -> dict | None:
    proxy = get_socks5_proxy()
    if not proxy:
        return None
    from urllib.parse import urlparse

    try:
        url = urlparse(_normalize_url(proxy))
        return {
            "host": url.hostname or "",
            "port": url.port or 0,
            "hasAuth": bool(url.username or url.password),
        }
    except Exception:
        return {"host": proxy, "port": 0, "hasAuth": False}
